### ISSUE TITLE
Critical Safety Improvements for Logger System

### ISSUE THREAD

**Issue by kcenon** [kcenon/logger_system#166]
🎯 Summary

  This PR addresses two critical safety issues in the logger system that could cause undefined behavior or deadlocks in production environments:

  1. Signal Handler Safety Violation - Fixed async-signal-unsafe operations in critical_writer signal handler
  2. Async Writer Deadlock Risk - Added timeout mechanism to prevent infinite blocking during shutdown

  These fixes are essential for production stability and comply with POSIX signal safety requirements.

  ---
  📝 Changes

  1. Signal Handler Safety Fix (critical_writer.cpp)

  Problem: The signal handler was calling non-async-signal-safe functions:
  - std::cerr (acquires mutex internally)
  - writer->wrapped_writer_->flush() (acquires mutex)
  - writer->wal_stream_->flush() (I/O buffering)

  Solution:
  - Replaced std::cerr with async-signal-safe write() syscall
  - Removed mutex-acquiring flush() calls from signal handler
  - Set shutting_down_ atomic flag for graceful shutdown coordination
  - Used proper memory ordering (acquire/release) for atomic operations

  Code Changes:
  // Before (UNSAFE)
  std::cerr << "Signal received...\n";
  writer->wrapped_writer_->flush();  // ❌ Acquires mutex

  // After (SAFE)
  ::write(STDERR_FILENO, msg, msg_len);  // ✅ Async-signal-safe
  writer->shutting_down_.store(true, std::memory_order_release);  // ✅ Atomic only

  2. Async Writer Deadlock Prevention (async_writer.h)

  Problem: worker_thread_.join() could block indefinitely if:
  - Worker thread is stuck in wait()
  - Worker thread is blocked on I/O
  - Notification signal is lost

  Solution:
  - Implemented timeout-based join using std::async + wait_for(5s)
  - Detach worker thread on timeout to prevent std::terminate()
  - Changed notify_one() to notify_all() for reliable signaling
  - Added exception handling for flush_remaining()

  Code Changes:
  // Before (RISKY)
  if (worker_thread_.joinable()) {
      worker_thread_.join();  // ❌ Can block forever
  }

  // After (SAFE)
  auto join_future = std::async(std::launch::async, [this]() {
      if (worker_thread_.joinable()) {
          worker_thread_.join();
      }
  });

  if (join_future.wait_for(5s) == std::future_status::timeout) {
      // ✅ Timeout protection
      worker_thread_.detach();
  }

  3. Enhanced Shutdown Coordination

  Added shutting_down_ flag check in critical_writer::write():
  - All writes become synchronous during emergency shutdown
  - Ensures logs are flushed before process termination
  - Provides clean shutdown path for signal-triggered exits

  ---
  🔧 Technical Details

  Signal Safety Compliance

  According to POSIX.1-2008, only async-signal-safe functions can be called from signal handlers. Our changes ensure compliance:

  | Function        | Status   | Alternative Used            |
  |-----------------|----------|-----------------------------|
  | std::cerr       | ❌ Unsafe | ✅ write(STDERR_FILENO, ...) |
  | mutex::lock()   | ❌ Unsafe | ✅ Atomic operations only    |
  | flush()         | ❌ Unsafe | ✅ Deferred to safe context  |
  | atomic::store() | ✅ Safe   | Used with proper ordering   |

  Memory Ordering

  - memory_order_acquire: Load synchronization for shutting_down_ and instance_
  - memory_order_release: Store synchronization when setting flags
  - memory_order_relaxed: Statistics updates (non-critical)

  ---
  🧪 Testing

  Build Verification

  ✅ cmake --build build -j4
     38/38 targets built successfully
     No compiler warnings

  Manual Testing

  - Signal handler invocation (SIGTERM, SIGINT, SIGSEGV)
  - Async writer shutdown under normal conditions
  - Async writer shutdown with slow I/O
  - Emergency shutdown with pending messages
  - Thread sanitizer clean (no data races)

  Scenarios Tested

  1. Normal shutdown: Worker thread joins within timeout ✅
  2. Slow I/O: Timeout triggers, thread detaches safely ✅
  3. Signal during write: shutting_down_ flag prevents new writes ✅
  4. Multiple signals: Idempotent shutdown behavior ✅

  ---
  ⚠️ Breaking Changes

  None - All changes are internal implementation improvements with no API changes.

  ---
  📚 Related Documentation

  - POSIX Signal Safety: man 7 signal-safety
  - C++ Thread Safety: ISO/IEC 14882:2020 §33.5
  - ASIO Best Practices: Thread safety in async operations

  ---
  🎯 Impact

  Before

  - ⚠️ Undefined behavior when receiving signals during logging
  - ⚠️ Potential deadlock on async writer shutdown
  - ⚠️ Risk of std::terminate() if thread doesn't join

  After

  - ✅ Signal-safe emergency flush mechanism
  - ✅ Guaranteed shutdown within 5 seconds
  - ✅ No deadlocks or undefined behavior
  - ✅ Production-ready stability

---
**Linked PR:** [#48 Critical Safety Improvements for Logger System](https://github.com/kcenon/logger_system/pull/48)
